libs/parallelism/execution/include/hpx/execution/sender_future.hpp

```diff
@@ -0,0 +1,137 @@
+            }
+
+            void set_done() noexcept
+            {
+                std::terminate();
+            }
+
+            template <typename U>
+            void set_value(U&& u) noexcept
+            {
+                data->set_value(std::forward<U>(u));
+                hpx::lcos::detail::intrusive_ptr_release(data);
+            }
+        };
+
+        template <typename Allocator>
+        struct make_future_receiver<void, Allocator>
+        {
+            hpx::lcos::detail::future_data_allocator<void, Allocator>* data;
+
+            void set_error(std::exception_ptr e) noexcept
+            {
+                data.set_exception(e);
+                hpx::lcos::detail::intrusive_ptr_release(data);
+            }
+
+            void set_done() noexcept
+            {
+                std::terminate();
+            }
+
+            template <typename Ts>
+            void set_value() noexcept
+            {
+                data.set_value(hpx::util::unused);
+                hpx::lcos::detail::intrusive_ptr_release(data);
+            }
+        };
+    }    // namespace detail
+
+    template <typename S, typename Allocator = hpx::util::internal_allocator<>,
+        typename = std::enable_if_t<is_sender_v<S>>>
+    auto make_future(S&& s, Allocator const& a = Allocator{})
+    {
+        using value_types =
+            typename hpx::execution::experimental::sender_traits<std::decay_t<
+                S>>::template value_types<hpx::util::pack, hpx::util::pack>;
+        // TODO: Don't use sync_wait_single_result (or rename it).
+        using result_type = std::decay_t<
+            typename detail::sync_wait_single_result<value_types>::type>;
+        using allocator_type = Allocator;
+        using shared_state_base =
+            hpx::lcos::detail::future_data_allocator<result_type,
+                allocator_type>;
+        using operation_state_type = typename hpx::util::invoke_result<
+            hpx::execution::experimental::connect_t, S,
+            detail::make_future_receiver<result_type, allocator_type>>::type;
+        using shared_state = detail::future_data<result_type, allocator_type,
+            operation_state_type>;
+        using other_allocator = typename std::allocator_traits<
+            allocator_type>::template rebind_alloc<shared_state>;
+        using allocator_traits = std::allocator_traits<other_allocator>;
+        using init_no_addref = typename shared_state::init_no_addref;
+        using unique_ptr = std::unique_ptr<shared_state,
+            util::allocator_deleter<other_allocator>>;
+
+        other_allocator alloc(a);
+        unique_ptr p(allocator_traits::allocate(alloc, 1),
+            hpx::util::allocator_deleter<other_allocator>{alloc});
+
+        allocator_traits::construct(alloc, p.get(), init_no_addref{}, alloc,
+            std::forward<S>(s),
+            detail::make_future_receiver<result_type, allocator_type>{p.get()});
```

Comment:
Shouldn't the `make_future_receiver` keep the shared state alive? If the returned future goes out of scope before the result has been received, the pointer to the shared state would have possibly become invalid. 

Hmmm. OTOH, this would create a cyclic dependency of the shared state to itself as the `make_future_receiver` is embedded in the new shared state, Can the operation_state break this cycle once it is triggered?

Reply:
I saw just now that you already break the cycle in the `make_future_receiver` itself - so all what's needed is another addref (possibly pass `true` as the second argument to the `future_access`). The shared state in the returned future should have a reference `count == 2` (one from the future and one from the `make_future_receiver`).

Reply:
Yes, the cycle is a bit annoying, but there are similar situations in some of the other senders as well and I think the base class works reasonably well. That's why I went with the promise/future pair initially because it was the easy solution.

I'll check add some sanity checks to the reference counts.

Reply:
I think the reference counts are a bit simpler now. The `init_no_addref` initializes the shared state to have a reference count of one (I think `init_no_addref` is very confusingly named), and then the intrusive pointer in the receiver increases it to two, meaning if I have the correct idea about how things are working the final access does not need to increase the reference count further.

Reply:
`init_no_addref` is confusing, I agree. Any suggestions for a better name (I'd be happy to change that)?

Reply:
I don't know if `init_addref` is any better? Basically `init_no_addref` tells me exactly the opposite of what it does, but that's because I didn't know first what it's referring to.

Reply:
It meant to refer to the fact that the new object does _not_ need any addref when assigned to the `intrusive_ptr` eventually (which doesn't make too much sense now that I speak out loudly about it). I will change it to `init_addref`, thanks!

Reply:
Yep, in your defence there is a comment that explains exactly that, so I don't mind either way. If you do end up changing the name please update the comment in that case as well ;)